core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -545,15 +546,39 @@ class Partition(val topic: String,
     laggingReplicas
   }
 
-  def appendRecordsToFutureReplica(records: MemoryRecords) {
-    getReplica(Request.FutureLocalReplicaId).get.log.get.appendAsFollower(records)
+  private def doAppendRecordsToFollower(records: MemoryRecords, isFuture: Boolean): Unit = {
+      if (isFuture)
+        getReplica(Request.FutureLocalReplicaId).get.log.get.appendAsFollower(records)
+      else {
+        // The read lock is needed to prevent the follower replica from being updated while ReplicaAlterDirThread
+        // is executing maybeDeleteAndSwapFutureReplica() to replace follower replica with the future replica.
+        inReadLock(leaderIsrUpdateLock) {
+           getReplica().get.log.get.appendAsFollower(records)
+        }
+      }
   }
 
-  def appendRecordsToFollower(records: MemoryRecords) {
-    // The read lock is needed to prevent the follower replica from being updated while ReplicaAlterDirThread
-    // is executing maybeDeleteAndSwapFutureReplica() to replace follower replica with the future replica.
-    inReadLock(leaderIsrUpdateLock) {
-      getReplica().get.log.get.appendAsFollower(records)
+  def appendRecordsToFollower(records: MemoryRecords, isFuture: Boolean) {
+    try {
+      doAppendRecordsToFollower(records, isFuture)
+    } catch {
+      case e: UnexpectedAppendOffsetException =>
+        val replica = if (isFuture) getReplica(Request.FutureLocalReplicaId).get else getReplica().get
+        if (replica.logEndOffset.messageOffset == replica.logStartOffset) {
```

Comment:
Do you think we should validate that the unexpected records at least contain the expected log start offset?

Reply:
Would that require uncompressing and looking at all the records? What do you think about checking if an expected offset is in between first and last offset of records?